searx/engines/mullvad-leta.py

```diff
@@ -0,0 +1,320 @@
+from searx.enginelib.traits import EngineTraits
+from httpx import Response
+from lxml import html
+
+traits: EngineTraits
+
+search_url = "https://leta.mullvad.net"
+
+# about
+about = {
+    "website": search_url,
+    "wikidata_id": 'Q47008412',  # the Mullvad id - not leta, but related
+    "official_api_documentation": 'https://leta.mullvad.net/faq',
+    "use_official_api": False,
+    "require_api_key": False,
+    "results": 'HTML',
+}
+
+# engine dependent config
+categories = ['general', 'web']
+paging = True
+max_page = 50
+time_range_support = True
+time_range_dict = {
+    "day": "d1",
+    "week": "w1",
+    "month": "m1",
+    "year": "y1",
+}
+
+
+def request(query: str, params: dict):
+    '''pre-request callback
+    params<dict>:
+      method  : POST/GET
+      headers : {}
+      data    : {} # if method == POST
+      url     : ''
+      category: 'search category'
+      pageno  : 1 # number of the requested page
+    '''
+    params['url'] = search_url
+    params['method'] = 'POST'
+    params['data'] = {
+        "q": query,
+        "gl": "",  # TODO: Get country from engine traits
+        "oc": "on",  # TODO: get only-cache from engine traits
+        "start": "11",
+    }
+
+    if params['time_range'] in time_range_dict:
+        params['dateRestrict'] = time_range_dict[params['time_range']]
+    else:
+        params['dateRestrict'] = ''
+
+    if params['pageno'] > 1:
+        #  Page 1 is n/a, Page 2 is 11, page 3 is 21, ...
+        params['data']['start'] = ''.join([str(params['pageno'] - 1), "1"])
+    return params
+
+
+def response(resp: Response):
+    '''post-response callback
+    resp: requests response object
+    '''
+    def extract_result(dom_result: html.HtmlElement):
+        [a_elem, h3_elem, p_elem] = dom_result.xpath(
+            'div/div')[0].getchildren()
+        return {
+            'url': a_elem.text,
+            'title': h3_elem.text_content(),
+            'content': p_elem.text,
+        }
+    dom = html.fromstring(resp.text)
+    search_results = dom.body.xpath('//main/div[2]')[0].getchildren()[1:]
+    return [extract_result(sr) for sr in search_results]
+
+
+# List of all available countries for the "gl" option
+_countries = [
```

Comment:
Oh, I just saw they're part of https://github.com/searxng/searxng/pull/3103/files#diff-fe29dd443723cfce49c4b43476d6311cfb8970e2515dc80955d38e15363a57bd. So we should probably fetch them dynamically from the URL used there instead of hard coding.

Reply:
Yes, this is part of the TODO to fetch them with engine traits